working_villagers/jobs/craftsman.lua

```diff
@@ -0,0 +1,390 @@
+	end
+	assert(target_x == nil)
+	assert(target_y == nil)
+
+	local count = 0
+	for _,dep in pairs(recipe) do
+		count = count + recipe_requires(dep, item_name, nil, nil)
+	end
+	return count
+end
+
+-- called by take_func for all iterations
+function craft_tables.get_craftingsupplies(self, item_name, iteration, target_x, target_y)
+	assert(item_name ~= nil)
+	assert(iteration ~= nil)
+	local recipe = recipes[iteration]
+	assert(recipe ~= nil)
+	local count = recipe_requires(recipe, item_name, target_x, target_y)
+	if count == 0 then return nil end
+	return count
+end
+-- called by put_func for all iterations
+function craft_tables.is_craftingsupplies(self, item_name, iteration, target_x, target_y)
+	assert(item_name ~= nil)
+	assert(iteration ~= nil)
+	-- TODO 
+  local data = craft_tables.get_craftingsupplies(self, item_name, iteration, target_x, target_y);
+  return data ~= nil
+end
+
+function craft_tables.get_craft_table(item_name)
+	assert(item_name ~= nil)
+	for key, value in pairs(craft_tables.names) do
+		if item_name==key then
+			return value
+		end
+	end
+	return nil
+end
+
+function craft_tables.is_craft_table(item_name)
+	assert(item_name ~= nil)
+	local data = craft_tables.get_craft_table(item_name);
+	return data ~= nil
+end
+
+local function find_craft_table_node(pos)
+	assert(pos       ~= nil)
+	local node = minetest.get_node(pos);
+	local data = craft_tables.get_craft_table(node.name);
+	return data ~= nil
+end
+
+local searching_range = {x = 10, y = 3, z = 10}
+
+local function put_func(villager,stack,data)
+	assert(villager  ~= nil)
+	assert(stack     ~= nil)
+
+	local target_x
+	if data == nil then target_x = nil else target_x = data.target_x end
+	local target_y
+	if data == nil then target_y = nil else target_y = data.target_y end
+
+	if data ~= nil and data.iteration ~= nil then
+		return not craft_tables.is_craftingsupplies(villager,stack:get_name(), data.iteration, target_x, target_y)
+	end
+
+	local ntarget = #recipes
+	for iteration=1,ntarget,1 do
+		if craft_tables.is_craftingsupplies(villager,stack:get_name(), iteration, target_x, target_y) then
+			return false
+		end
+	end
+	return true;
+end
+local function take_func(villager,stack,data)
+	assert(villager  ~= nil)
+	assert(stack     ~= nil)
+	assert(data      == nil)
+	local item_name = stack:get_name()
+	local ntarget = #recipes
+	for iteration=1,ntarget,1 do
+		if craft_tables.is_craftingsupplies(villager,item_name, iteration, nil, nil) then
+			local inv = villager:get_inventory()
+			local itemstack = ItemStack(item_name)
+			local count = craft_tables.get_craftingsupplies(villager,item_name, iteration, nil, nil)
+			assert(count ~= nil)
+			assert(count ~= 0)
+			-- TODO allow to specify number of copies to make
+			itemstack:set_count(craft_tables.get_craftingsupplies(villager,item_name, iteration, nil, nil))
+			if (not inv:contains_item("main", itemstack)) then
+				return true
+			end
+		end
+	end
+	return false
+end
+
+
+
+
+
+
```

Comment:
There is a lot of empty space here. Can that be removed?

Reply:
I cleaned things up a bit. Also reorganized things for consistency.